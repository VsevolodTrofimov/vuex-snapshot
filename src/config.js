/**
 * @namespace 
 * @property {Boolean} autoResolve resolve all MockPromises and fetches in order they were created
 * @property {Boolean} snapEnv include state, getters and paylaod into snapshot
 */
const options = {
  autoResolve: false,
  snapEnv: false
}

// they are likely to stay flat
const defaults = Object.assign({}, options)
const reset = () => Object.assign(options, defaults)
reset()

export default {
  options,
  reset
}