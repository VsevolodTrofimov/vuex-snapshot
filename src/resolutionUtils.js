import serialize from './serialize'


export const normalizeResolution = resolution => {
  const normalResolution = {
    name: '',
    type: 'resolve',
    payload: undefined,
  }
  
  // string constructor
  if(typeof resolution === 'string' || resolution.indexOf('string')) {
    normalResolution.name = resolution
    return normalResolution
  } 
  
  // object constructor, errors are duplicated because they are likely to occur in promises
  if(resolution.name) {
    normalResolution.name = resolution.name
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution must have a name')
  }

  if(resolution.type && ['resolve', 'reject'].indexOf(resolution.type) !== -1) {
    normalResolution.type = resolution.type
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution type must be' 
                    + 'either "resovle" or "reject"')
  }

  if(resolution.payload) {
    normalResolution.payload = resolution.payload
  }
}


export const simualteResolution = (resolution, snapCb, timetable) => {
  snapCb(`RESOLUTION: ${resolution.name} -> ${resolution.type}`, resolution.payload)
  return timetable.trigger(resolution)
}


export const simualteResolutions = (resolutions, snapCb, timetable) => {
  return new Promise((resolveSimulation, rejectSimulation) => {
    const normalResolutions = resolutions.map(normalizeResolution)
    
    // simulates given resolution and queues the next until all are simulated
    const simulationLoop = idx => {
      if(idx === normalResolutions.length) {
        resolveSimulation()
      }
      else {
        simualteResolution(normalResolutions[idx], snapCb, timetable)
          .then(() => simulationLoop(idx + 1))
          .catch(rejectSimulation)
      }
    }

    simulationLoop(0)
  })
}
