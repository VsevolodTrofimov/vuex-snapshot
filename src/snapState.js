import serialize from './serialize'


/**
 * Serializes your state the fancy way 
 * (including functions and mock promises)
 * @param {*} state 
 */
export const snapState = state => serialize(state)
export default snapState