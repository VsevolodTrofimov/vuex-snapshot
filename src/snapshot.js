class Snapshot {
  constructor() {
    this.value = []
    this.add = this.add.bind(this)
  }
  add(message, payload) {
    const entry = {}
    entry.message = message
    if(typeof payload !== 'undefined') entry.payload = payload
    this.value.push(entry)
  }
}


export default Snapshot