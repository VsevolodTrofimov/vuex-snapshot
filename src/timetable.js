import {makeSuffix, find} from './utility'


const RealPromise = Promise
export const entries = []


export const register = ({name, promise, payload, resolve, reject}) => {
  const entry = {
    payload,
    promise,
    resolve,
    reject,
    called: false
  }

  // make sure name is unique
  const suffix = makeSuffix()
  while(find(entries, e => e.name === name + suffix())) suffix.next()

  entry.name = name + suffix()

  entries.push(entry)
}


export const trigger = ({name, type, payload}) => {
  const suffix = makeSuffix()
  while(find(entries, e => e.name === name + suffix() && e.called)) suffix.next()
  
  const entry = find(entries, e => e.name === name + suffix())
  
  return new RealPromise((resolve, reject) => {
    if(typeof entry === 'undefined') {
      reject(new Error(`vuex-snapshot: did not find ${name + suffix()} that wasn't already resolved or rejected`))
    }
  
    entry[type](payload)
    entry.called = true
    entry.promise
      .then(resolve)
      .catch(resolve)
  })
}


export const ensureAbsence = (promise) => {
  for(let i = 0; i < entries.length; ++i) {
    if(entries[i].promise === promise) {
      entries.splice(i, 1)
    }
  }
}

export const reset = () => entries.length = 0

export default {
  register,
  trigger,
  reset,
  entries,
  ensureAbsence,
}