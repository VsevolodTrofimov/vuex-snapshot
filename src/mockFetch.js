import timetable from './timetable'
import useGlobally from './useGlobally'


const RealPromise = Promise
const realFetch = fetch


// So we can resolve them manually and serialize them
export const mockFetch = (url, init) => {
  let resovleTrigger
  let rejectTrigger

  const cbProxy = (resolve, reject) => {
    resovleTrigger = resolve
    rejectTrigger = reject
  }

  const simulation = new RealPromise(cbProxy)

  timetable.register({
    name: url,
    payload: init,
    resolve: resovleTrigger,
    reject: rejectTrigger
  })

  return simulation
}


export const useMock = () => useGlobally('fetch', mockFetch)
export const useReal = () => useGlobally('fetch', realFetch)