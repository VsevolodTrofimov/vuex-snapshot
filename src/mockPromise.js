import timetable from './timetable'
import {useGlobally} from './utility'


const RealPromise = Promise


export class MockPromise extends RealPromise {
  /**
   * Named promise that can be resolved manually and properly serialized 
   * and registers it in timetable
   * @param {Function} cb 
   * @param {string} name
   */
  constructor(cb, name='Promise') {
    let resovleTrigger
    let rejectTrigger

    // name-only construction
    if(typeof cb === 'string') {
      name = cb
      cb = () => {}
    }

    const cbProxy = (resolve, reject) => {
      resovleTrigger = resolve
      rejectTrigger = reject
      cb(resolve, reject)
    }

    super(cbProxy)

    this.name = name
    this.resolve = resovleTrigger
    this.reject = rejectTrigger

    timetable.register({
      name,
      promise: this,
      payload: cb,
      resolve: resovleTrigger,
      reject: rejectTrigger
    })
  }
}


export const useMock = () => useGlobally('Promise', MockPromise)
export const useReal = () => useGlobally('Promise', RealPromise)