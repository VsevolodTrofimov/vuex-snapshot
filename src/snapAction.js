import Snapshot from './snapshot'
import {simualteResolutions} from './resolutionUtils'
import timetable from '../src/timetable'

const RealPromise = Promise


export const makeCallSnapper = (snapshot, type, cb) => (name, payload) => {
  snapshot.add(`${type}: ${name}`, payload)
  cb(name, payload)
}


/**
 * @typedef {{name:string, type: ("resolve" | "reject"), payload}} Resolution
 */
/**
 * Takes snapshot of action's evaluation
 * @param {Function} action action to test
 * @param {{state, getters, commit: Function, dispatch: Function, payload}} mocks arguments passed to the action, payload is the second argument
 * @param {[(string | Resolution)]} resolutions
 * @param {{autoResovle: Boolean, snapEnv: Boolean}} options
 * @returns  {(string | Promise<string>)}
 */
export const snapAction = (action, mocks, resolutions, options) => {
  const snapshot = new Snapshot()
  const mockCommit = makeCallSnapper(snapshot, 'COMMIT', mocks.commit)
  const mockDispatch = makeCallSnapper(snapshot, 'DISPATCH', mocks.dispatch)

  if(options.snapEnv) {
    snapshot.add('DATA MOCKS', {
      state: mocks.state,
      getters: mocks.getters
    })
    snapshot.add('ACTION CALL', mocks.payload)
  }

  const actionReturn = action({
    commit: mockCommit,
    dispatch: mockDispatch,
    state: mocks.state,
    getters: mocks.getters
  }, mocks.payload)

  if(typeof actionReturn !== 'undefined' && actionReturn instanceof Promise) {
    // action is async
    return new RealPromise((resolve, reject) => {
      actionReturn
        .then(payload => {
          snapshot.add('ACTION RESOLVED', payload)
          resolve(snapshot.value)
        })
        .catch(payload => {
          snapshot.add('ACTION REJECTED', payload)
          resolve(snapshot.value)
        })
      
      simualteResolutions(resolutions, snapshot, timetable, options)
        .then(() => {
          // this is needed to let action to resolve first
          setTimeout(() => {
            snapshot.add('ACTION DID NOT RESOLVE')
            resolve(snapshot.value)
          }, 0)
        })
        .catch(err => {
          reject({
            err,
            run: snapshot.value
          })
        })

    })
  } else {
    // action is sync
    return snapshot.value
  }
}

export default snapAction