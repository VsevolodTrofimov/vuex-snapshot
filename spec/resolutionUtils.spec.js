import {normalizeResolution, simualteResolution, simualteResolutions, lib} from '../src/resolutionUtils' 
import { MockPromise } from '../src/mockPromise'


describe('resolutionUtils', () => {
  describe('normalizeResolution', () => {
    it('Converts string into "resolve" resolution', () => {
      const resolution = normalizeResolution('name')
      expect(resolution.name).toBe('name')
      expect(resolution.type).toBe('resolve')

      const resolution2 = normalizeResolution(new String('name2'))
      expect(resolution2.name).toEqual('name2')
      expect(resolution2.type).toBe('resolve')
    })

    it('Throws when resolution has no name', () => {
      expect(() => {
        normalizeResolution({type: 'reject'})
      }).toThrow()

      expect(() => {
        normalizeResolution({})
      }).toThrow()
    })

    it('Throws when resolution type is nor resolve nor reject', () => {
      expect(() => {
        normalizeResolution({name: '1', type: 'fail'})
      }).toThrow()
      
      expect(() => {
        normalizeResolution({name: '1', type: 'ok'})
      }).toThrow()
    })

    it('Passes payload', () => {
      const payload = {}
      const resolution = normalizeResolution({name: '1', payload})
      expect(resolution.payload).toBe(payload)

      const resolutionZero = normalizeResolution({name: '1', payload: 0})
      expect(resolution.payload).toBe(payload)
    })

    it('Uses resolve as a default type', () => {
      const resolution = normalizeResolution({name: '1', payload: 0})
      expect(resolution.type).toBe('resolve')
    })
  })

  
  describe('Simualte resolution', () => {
    let testResolution
    const snapshot = {}
    const timetable = {}

    beforeEach(() => {
      testResolution = {
        name: 'someName',
        type: 'resolve',
        payload: {a: 'b'}
      }

      snapshot.add = jest.fn()
      timetable.trigger = jest.fn(() => Promise.resolve())
    })

    it('Returns timetable trigger result', () => {
      const timetableReturn = Promise.resolve()
      timetable.trigger.mockReturnValue(timetableReturn)

      const simutationReturn = simualteResolution(testResolution, snapshot, timetable)
      expect(simutationReturn).toBe(timetableReturn)
    })

    it('Produces message that matches snapshot', () => {
      simualteResolution(testResolution, snapshot, timetable)
      expect(snapshot.add.mock.calls[0]).toMatchSnapshot()
    })
  })

  
  describe('Simualte resolutions', () => {
    let testResolution
    let resolutions
    let normalizeResolutionSpy, simualteResolutionSpy
    const snapshot = {}, timetable = {}, options = {}

    beforeEach(() => {
      testResolution = {
        name: 'someName',
        type: 'reject',
        payload: {a: 'b'}
      }

      resolutions = [testResolution, 'a', 'b']

      snapshot.add = jest.fn()
      timetable.trigger = jest.fn().mockReturnValue(Promise.resolve())

      lib.normalizeResolution = jest.fn(normalizeResolution)
      lib.simualteResolution = jest.fn().mockReturnValue(Promise.resolve())
    })
  
    it('Returns a promise', () => {
      const returned = simualteResolutions(resolutions, snapshot, timetable, options)
      expect(returned instanceof Promise).toBe(true)
    })

    it('Normalizes all resolution', done => {
      simualteResolutions(resolutions, snapshot, timetable, options)
        .then(() => {
          expect(
            lib.normalizeResolution.mock.calls.length
          ).toBe(resolutions.length)
          
          done()
        })
        .catch(done)
    })

    it('Simulates all resolutions', done => {
      simualteResolutions(resolutions, snapshot, timetable, options)
        .then(() => {
          expect(
            lib.simualteResolution.mock.calls.length
          ).toBe(resolutions.length)
          
          done()
        })
        .catch(done)
    })

    it('Simulates resolutions one by one', done => {
      const p1 = new MockPromise('p1')
      const p2 = new MockPromise('p2')
      const p3 = new MockPromise('p3')
    
      p1.then(() => {
        expect(timetable.trigger.mock.calls.length).toBe(1)
      })

      p2.then(() => {
        expect(timetable.trigger.mock.calls.length).toBe(1)
      })

      p3.then(() => {
        expect(timetable.trigger.mock.calls.length).toBe(1)
        return new MockPromise('p4')
      })

      simualteResolutions(['p1', 'p2', 'p3', 'p4'], snapshot, timetable, options)
        .then(() => {
          done()
        })
    })
    
    it('Pipes errors up', done => {
      lib.normalizeResolution = normalizeResolution
      lib.simualteResolution = simualteResolution

      const normalizeErrorPromise = new MockPromise(() => {})
      const snapshotErrorPromise = new MockPromise(() => {})
      const timetableErrorPromise = new MockPromise(() => {})

      simualteResolutions([{type: 'strange thing'}], snapshot, timetable, options)
        .then(normalizeErrorPromise.reject)
        .catch(err => {
          expect(err instanceof Error).toBe(true)
          normalizeErrorPromise.resolve()
        })

      simualteResolutions(resolutions, {}, timetable, options)
      .then(snapshotErrorPromise.reject)
        .catch(err => {
          expect(err instanceof Error).toBe(true)
          snapshotErrorPromise.resolve()
        })

      simualteResolutions(resolutions, {}, timetable, options)
        .then(timetableErrorPromise.reject)
        .catch(err => {
          expect(err instanceof Error).toBe(true)
          timetableErrorPromise.resolve()
        })
        
      Promise.all([
        normalizeErrorPromise, 
        snapshotErrorPromise,
        timetableErrorPromise,
      ])
        .then(() => {
          done()
        })
        .catch(done)
    })
  })
})