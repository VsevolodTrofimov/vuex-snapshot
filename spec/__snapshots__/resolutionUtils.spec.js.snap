// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolutionUtils Simualte resolution Produces message that matches snapshot 1`] = `
Array [
  "RESOLUTION: someName -> resolve",
  Object {
    "a": "b",
  },
]
`;
